stdlib/public/core/StringWordBreaking.swift

```diff
@@ -0,0 +1,679 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension _StringGuts {
+  internal func roundDownToNearestWord(
+    _ i: String.Index
+  ) -> String.Index {
+    _internalInvariant(i._encodedOffset <= count)
+
+    let offset = i._encodedOffset
+
+    if offset == 0 || offset == count {
+      return i
+    }
+
+    return _slowRoundDownToNearestWord(i)
+  }
+
+  @inline(never)
+  internal func _slowRoundDownToNearestWord(
+    _ i: String.Index
+  ) -> String.Index {
+    let offset = i._encodedOffset
+    let start = previousWordIndex(endingAt: offset)
+    let end = nextWordIndex(startingAt: start)
+    _internalInvariant(offset <= end, "Word breaking inconsistency")
+
+    if offset == end {
+      return i
+    }
+
+    return String.Index(_encodedOffset: start)
+  }
+
+  @inline(never)
+  @_effects(releasenone)
+  internal func nextWordIndex(startingAt i: Int) -> Int {
+    if _slowPath(isForeign) {
+      return _foreignNextWordIndex(startingAt: i)
+    }
+
+    return withFastUTF8 { utf8 in
+      nextWordBoundary(startingAt: i) {
+        _internalInvariant($0 >= 0)
+
+        guard $0 < utf8.count else {
+          return nil
+        }
+
+        let (scalar, len) = _decodeScalar(utf8, startingAt: $0)
+        return (scalar, $0 &+ len)
+      }
+    }
+  }
+
+  internal func _foreignNextWordIndex(startingAt i: Int) -> Int {
+#if _runtime(_ObjC)
+    return nextWordBoundary(startingAt: i) {
+      _internalInvariant($0 >= 0)
+
+      guard $0 < count else {
+        return nil
+      }
+
+      let scalars = String.UnicodeScalarView(self)
+      let idx = String.Index(_encodedOffset: $0)
+
+      let scalar = scalars[idx]
+      let nextIndex = scalars.index(after: idx)
+
+      return (scalar, nextIndex._encodedOffset)
+    }
+#else
+    fatalError("No foreign strings on this platform in this version of Swift.")
+#endif
+  }
+
+  internal func previousWordIndex(endingAt i: Int) -> Int {
+    if _slowPath(isForeign) {
+      return _foreignPreviousWordIndex(endingAt: i)
+    }
+
+    return withFastUTF8 { utf8 in
+      previousWordBoundary(endingAt: i) {
+        _internalInvariant($0 <= count)
+
+        guard $0 > 0 else {
+          return nil
+        }
+
+        let (scalar, len) = _decodeScalar(utf8, endingAt: $0)
+        return (scalar, $0 &- len)
+      }
+    }
+  }
+
+  @inline(never)
+  internal func _foreignPreviousWordIndex(endingAt i: Int) -> Int {
+    #if _runtime(_ObjC)
```

Comment:
Nit: misaligned indentation

```suggestion
#if _runtime(_ObjC)
```